tools/eslint-plugin/dist/rules/no-internal-barrel-imports.js

```diff
@@ -0,0 +1,297 @@
+    const ignoredBarrelModules = (
+      (maybeTsConfig &&
+        extraOpts &&
+        extraOpts[OPTION_IGNORED_BARREL_MODULES]) ||
+      []
+    ).map((p) => {
+      /** @type {string} earlier check means it is definitely a string */
+      const tsConfig = maybeTsConfig;
+      return OsPaths.normalize(OsPaths.resolve(OsPaths.dirname(tsConfig), p));
+    });
+
+    return {
+      ImportDeclaration(node) {
+        /** @param {ts.Symbol | undefined} symbol */
+        function getRelativeImportPathForExportedSymbol(symbol) {
+          if (symbol === undefined) return "";
+          const declaration = symbol.valueDeclaration || symbol.declarations[0];
+          const fileOfExport = declaration.getSourceFile();
+          const path = TsImportPaths.normalize(
+            withoutExt(
+              TsImportPaths.relative(
+                TsImportPaths.dirname(thisModule.fileName),
+                fileOfExport.fileName
+              )
+            )
+          );
+          // a path unprefixed by "./" is interpretted as from node_modules
+          const ensurePrefixed = (path) =>
+            path[0] === "." ? path : `./${path}`;
+          return ensurePrefixed(path);
+        }
+
+        if (typeof node.source.value !== "string")
+          throw Error("Invalid input source");
+
+        /** @type{ts.ImportDeclaration} */
+        const importNodeTs = parserServices.esTreeNodeToTSNodeMap.get(node);
+        if (!importNodeTs)
+          throw Error("equivalent typescript node could not be found");
+
+        if (importNodeTs.importClause && importNodeTs.importClause.isTypeOnly)
+          return;
+
+        const thisModule = importNodeTs.getSourceFile();
+
+        if (!thisModule.resolvedModules.has(importNodeTs.moduleSpecifier.text))
+          throw Error("module did not have this import");
+
+        const importInfo = thisModule.resolvedModules.get(
+          importNodeTs.moduleSpecifier.text
+        );
+
+        const importIsPackage =
+          importInfo === undefined || importInfo.isExternalLibraryImport;
+        if (importIsPackage) return;
+
+        // path case can be different between the cased import specifier, and the resolved system path
+        // use lower case on OSs known to use lower cased file systems (e.g. windows and mac but not linux)
+        const targetFilePathOfThisImportStmt =
+          process.platform === "win32" || process.platform === "darwin"
+            ? importInfo.resolvedFileName.toLowerCase()
+            : importInfo.resolvedFileName;
+
+        const importedModule = program.getSourceFileByPath(
+          targetFilePathOfThisImportStmt
+        );
+
+        if (!importedModule) throw Error("couldn't find imported module");
+
+        // prettier-ignore
+        if (ignoredBarrelModules.includes(OsPaths.normalize(importedModule.fileName)))
+          return;
+
+        /** @type {ts.ImportDeclaration["moduleSpecifier"][]} */
+        const imports = importedModule.imports;
```

Comment:
again, this is internal and may change at any time with a Typescript update, breaking your rule